packages/app-builder-lib/src/macPackager.ts

```diff
@@ -91,6 +91,58 @@ export default class MacPackager extends PlatformPackager<MacConfiguration> {
     }
   }
 
+  protected async doPack(outDir: string, appOutDir: string, platformName: ElectronPlatformName, arch: Arch, platformSpecificBuildOptions: MacConfiguration, targets: Array<Target>): Promise<any> {
+    switch (arch) {
+      default: {
+        return super.doPack(outDir, appOutDir, this.platform.nodeName as ElectronPlatformName, arch, this.platformSpecificBuildOptions, targets);
+      }
+      case Arch.universal: {
+        const x64Arch = Arch.x64;
+        const x64AppOutDir = appOutDir + '-' + Arch[x64Arch];
+        const arm64Arch = Arch.arm64;
+        const arm64AppOutPath = appOutDir + '-' + Arch[arm64Arch];
+        return super.doPack(outDir, x64AppOutDir, platformName, x64Arch, platformSpecificBuildOptions, targets, false).then(() => 
+          super.doPack(outDir, arm64AppOutPath, platformName, arm64Arch, platformSpecificBuildOptions, targets, false)).then(() => {
+            const framework = this.info.framework
+            log.info({
+              platform: platformName,
+              arch: Arch[arch],
+              [`${framework.name}`]: framework.version,
+              appOutDir: log.filePath(appOutDir),
+            }, `packaging`)
+            const appFile = `${this.appInfo.productFilename}.app`;
+            const { makeUniversalApp } = require('@electron/universal');
+              return makeUniversalApp({
+                x64AppPath: path.join(x64AppOutDir, appFile),
+                arm64AppPath: path.join(arm64AppOutPath, appFile),
+                outAppPath: path.join(appOutDir, appFile),
+              }).then(() => {
+                const rmdir = (path: string, options: any) => {
+                  return new Promise<void>((resolve, reject) => {
+                    const fs = require('fs');
+                    return fs.rmdir(path, options, (err: Error) => {
+                      if (err) {
+                        reject(err);
+                      }
+                      else {
+                        resolve();
+                      }
+                    });
+                  });
+                };
+                return Promise.all([
+                  rmdir(x64AppOutDir, { recursive: true }),
+                  rmdir(arm64AppOutPath, { recursive: true }),
+                ]).then(() => {
+                  return this.doSignAfterPack(outDir, appOutDir, platformName, arch, platformSpecificBuildOptions, targets);
+                });
+              });
+            });
+        break;
```

Comment:
That's not how this is being used in practice. Universal exists to avoid the existence of platform specific builds, not to bring them back in a hidden way or special case updaters. If an app builds universal, it would not build or ship x86 or arm64 targets.

Using a flag, where in the pipeline would the universal code run? It would ultimately have to do all the same things a regular MacPackager does and hide the existence of the x86 and arm64 build results. 

Is there a sample project or repro for the node-gyp-rebuild issue?

I can abandon this PR if you want to start a new one if there are clear answers but sounds like it isn't all that simple and more trading one set of problems for another?

Reply:
Sorry for the delay on this one.
You're completely correct that the approachs will be trading one set of problems for another. 😅 

Quick note: `electron-updater` currently expects to find an arch-specific build. 
https://github.com/electron-userland/electron-builder/blob/2b116d138e9effa3314fe6ff4474b639f7fddf47/packages/electron-updater/src/MacUpdater.ts#L34-L35
After much time for users to upgrade to M1, the universal binaries will be dropped anyhow, and builds will just need to include arm64, which electron-updater will already be ready to handle.